data/transactions/logic/eval.go

```diff
@@ -992,6 +1006,74 @@ func (err EvalError) Unwrap() error {
 	return err.Err
 }
 
+func (cx *EvalContext) evalError(err error) error {
+	pc, det := cx.pcDetails()
+	details := fmt.Sprintf("pc=%d, opcodes=%s", pc, det)
+
+	return EvalError{
+		serr.Annotate(err, "pc", pc, "group-index", cx.groupIndex,
+			"eval-states", cx.evalStates()),
+		details,
+		cx.runMode == ModeSig,
+	}
+}
+
+type evalState struct {
+	Scratch []any    `json:"scratch,omitempty"`
+	Stack   []any    `json:"stack,omitempty"`
+	Logs    [][]byte `json:"logs,omitempty"`
+}
+
+func (cx *EvalContext) evalStates() []evalState {
+	states := make([]evalState, cx.groupIndex+1)
+	for i := 0; i <= cx.groupIndex; i++ {
+		var scratch []stackValue
+		if cx.pastScratch[i] != nil {
+			scratch = (*cx.pastScratch[i])[:]
+		}
+		lastNonZero := -1
+		scratchAsAny := make([]any, len(scratch))
+		for s, sv := range scratch {
+			if !sv.isEmpty() {
+				lastNonZero = s
+			}
+			scratchAsAny[s] = sv.asAny()
+		}
+		if lastNonZero == -1 {
+			scratchAsAny = nil
+		} else {
+			scratchAsAny = scratchAsAny[:lastNonZero+1]
+		}
+
+		// Only the current program's stack is still available. So perhaps it
+		// should be located outside of the evalState, with the PC.
+		var stack []any
+		if cx.groupIndex == i {
+			stack = convertSlice(cx.Stack, func(sv stackValue) any {
+				return sv.asAny()
+			})
+		}
+
+		states[i] = evalState{
+			Scratch: scratchAsAny,
+			Stack:   stack,
+			Logs:    convertSlice(cx.TxnGroup[i].EvalDelta.Logs, func(s string) []byte { return []byte(s) }),
+		}
+	}
+	return states
+}
+
+func convertSlice[X any, Y any](input []X, fn func(X) Y) []Y {
```

Comment:
nit: mapSlice?

Reply:
This function exists with this name in our REST code.  I do like `mapSlice` better, but thought it would be nicer to have consistency. (In the end, I imagine this function ought to exist in a small, common library, once something wins out in the Go ecosystem.)

I do also see some argument against using the word map in the Go world, since it's so ingrained as meaning the map data structure.